sdb/commands/linux/ps.py

```diff
@@ -0,0 +1,180 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import datetime
+import argparse
+from textwrap import shorten
+from typing import Callable, Dict, Iterable, Union
+
+import drgn
+from drgn.helpers.linux.pid import for_each_task
+from drgn.helpers.linux.mm import cmdline
+
+import sdb
+from sdb.commands.internal.table import Table
+from sdb.commands.stacks import Stacks
+
+
+def _cmdline(obj: drgn.Object) -> str:
+    try:
+        s = " ".join(map(lambda s: s.decode("utf-8"), cmdline(obj)))
+
+        #
+        # The command line for a given thread can be obnoxiously long,
+        # so (by default) we limit it to 50 characters here. This helps
+        # preserve the readability of the command's output, but comes at
+        # the cost of not always showing the full command line of a
+        # thread.
+        #
+        return shorten(s, width=50)
+    except drgn.FaultError:
+        #
+        # The command line information is contained in the user address
+        # space of each thread, rather than in the kernel's address
+        # space. Thus, often, it may not be possible to retreive the
+        # thread's command line; e.g. when reading from a core dump.
+        #
+        return ""
+
+
+class Ps(sdb.Locator, sdb.PrettyPrinter):
+    """
+    Locate and print information about processes
+
+    POTENTIAL COLUMNS
+        task - address of the task_struct
+        uid  - user id
+        pid  - the pid of the thread's process
+        time - cumulative CPU time, "[DD-]HH:MM:SS" format
+        state - the state of the thread
+        prio - the priority of the thread
+        comm - the thread's command
+        cmdline - the thread's command line (when available)
+
+    EXAMPLE
+        sdb> ps -e
+        TASK               UID  PID  TIME    STIME   PPID  CMD
+        ------------------ ---- ---- ------- ------- ----- ----------------
+        0xffff9995001a0000 1000 4387 0:00:00 0:00:00 1218  inotify_reader
+        0xffff9995001a2d00 1000 4382 0:00:00 0:00:00 1218  slack
+        0xffff9995001a4380 1000 4383 0:00:04 0:00:00 1218  slack
+        0xffff9995001a5a00 1000 4554 0:11:00 0:09:56 4118  AudioIP~ent RPC
+        0xffff99950a430000 1000 4284 0:00:00 0:00:00 4118  HTML5 Parser
+        0xffff99950a431680 1000 4157 0:00:12 0:00:08 1218  localStorage DBa
+        ...
+    """
+
+    names = ["ps"]
+    input_type = "struct task_struct *"
+    output_type = "struct task_struct *"
+
+    FIELDS: Dict[str, Callable[[drgn.Object], Union[str, int]]] = {
```

Comment:
Not that it matters if we go ahead with layering the `ps` command on top of `threads` but I don't see `task` mentioned here. What if a user wants to specify it with the `-o` option?

Reply:
updated.